### PR TITLE
miner: log locally mined blocks after they are 5-deep in the chain

### PR DESCRIPTION
**Pull request by carver**
For issue #1047

Please review structure and go idioms.  I'm a go newb!

Local deep mining logs help determine which blocks are unlikely to end up as uncles.

Mechanism:
 * Store the 5 most recent locally mined block numbers
 * On every imported block, check if the 5-deep block num is in that store
 * Also confirm that the block is signed with miner's coinbase
 * If so, write Info log

Why not just check the coinbase? This log is useful if you're running
multiple nodes and want to know if miners farming for *this* node are performing well.

### CODE REVIEW

**Comment by carver** [comment]
Hm, I think I busted something.  When I turn the ethminer on, I get these logs in geth:
```
I0521 17:38:44.147453    2926 miner.go:87] Can not start mining operation due to network sync (starts when finished)
I0521 17:38:44.648235    2926 miner.go:87] Can not start mining operation due to network sync (starts when finished)
I0521 17:38:45.148975    2926 miner.go:87] Can not start mining operation due to network sync (starts when finished)
I0521 17:38:45.649728    2926 miner.go:87] Can not start mining operation due to network sync (starts when finished)
I0521 17:38:46.150510    2926 miner.go:87] Can not start mining operation due to network sync (starts when finished)
I0521 17:38:46.651276    2926 miner.go:87] Can not start mining operation due to network sync (starts when finished)
I0521 17:38:47.152081    2926 miner.go:87] Can not start mining operation due to network sync (starts when finished)
I0521 17:38:47.652907    2926 miner.go:87] Can not start mining operation due to network sync (starts when finished)
I0521 17:38:48.153636    2926 miner.go:87] Can not start mining operation due to network sync (starts when finished)
I0521 17:38:48.654400    2926 miner.go:87] Can not start mining operation due to network sync (starts when finished)
I0521 17:38:49.155157    2926 miner.go:87] Can not start mining operation due to network sync (starts when finished)
I0521 17:38:49.655902    2926 miner.go:87] Can not start mining operation due to network sync (starts when finished)
I0521 17:38:50.156679    2926 miner.go:87] Can not start mining operation due to network sync (starts when finished)
```
and these in ethminer:
```
  ℹ  17:38:37|ethminer  Getting work package...
  ℹ  17:38:43|ethminer  Loading from libethash...
  ℹ  17:38:43|ethminer  Done loading.
  ℹ  17:38:44|ethminer  Getting work package...
  ℹ  17:38:44|ethminer  Getting work package...
  ℹ  17:38:45|ethminer  Getting work package...
  ℹ  17:38:45|ethminer  Getting work package...
  ℹ  17:38:46|ethminer  Getting work package...
  ℹ  17:38:46|ethminer  Getting work package...
  ℹ  17:38:47|ethminer  Getting work package...
  ℹ  17:38:47|ethminer  Getting work package...
```
Any ideas?

**Comment by karalabe** [comment]
I believe the miner got broken in latest develop due to a trivial mistake. Supposedly it's fixed by https://github.com/ethereum/go-ethereum/commit/054abe20b81ada11a820405f4a42f59cea9f9199. Please try to pull in the latest dev (8 min fresh) and try again.

**Comment by carver** [comment]
My live test consists of:
- [x] confirming that a mined block accepted by the blockchain does produce the "5 deep" message
- [x] confirming that a mined block uncle does not produce the "5 deep" message

(where "found on the chain" is verified with http://etherapps.info )

Functionally, this is ready to merge.

**Comment by obscuren** [comment]
Please get rid of https://github.com/carver/go-ethereum/commit/6e77ca7a0eae637e3ef74283833736ab3dd81610 and rebase instead of merging :-)

**Comment by obscuren** [comment]
Regarding the miner being locked, that was indeed a mistake. Should be fixed

**Comment by carver** [comment]
Cool, pushed the rebase

**Comment by carver** [comment]
I don't have OSX to test on, can someone help out?

File: miner/worker.go [hunk 1]
```diff
@@ -38,6 +38,13 @@ type Agent interface {
 	GetHashRate() int64
 }
 
+const MINING_LOG_AT_DEPTH = 5
+
+type UInt64RingBuffer struct {
```
Comment (zelig) [inline miner/worker.go]: this should not be exported. Use lowercase

File: miner/worker.go [hunk 2]
```diff
@@ -38,6 +38,13 @@ type Agent interface {
 	GetHashRate() int64
 }
 
+const MINING_LOG_AT_DEPTH = 5
```
Comment (zelig) [inline miner/worker.go]: Not sure this should be exported. Either way, we also use camel convention for constants. Please use `miningLogAtDepth`

**Comment by carver** [comment]
@zelig Thanks, done

**Comment by obscuren** [comment]
It doesn't appear to be working properly (no output) :(

**Comment by carver** [comment]
Bummer.  How are you testing it?  I got log output when mining the live Olympic network on my Ubuntu box.

**Comment by obscuren** [comment]
A local chain with `geth --datadir /tmp/001 --networkid 1234` and start mining

**Comment by carver** [comment]
Thanks, I will replicate and push a fix

**Comment by carver** [comment]
The code wasn't handling the case of all blocks mined.  Two issues resolved.

I see the deep log messages now on a local network, where you are the only miner.

**Pull request merged**